File: src_nextgen/actool_linux/paletteimg.py
from __future__ import annotations

from dataclasses import dataclass
import struct


MAGIC = 0xCAFEF00D
MAX_PALETTE_COLORS = 0x1000


@dataclass(frozen=True)
class ThemePixelRendition:
    version: int
    compression_type: int
    raw_data: bytes


@dataclass(frozen=True)
class QuantizedImageData:
    version: int
    palette: tuple[bytes, ...]  # ARGB entries (4 bytes each)
    indices: bytes              # one byte per pixel after unpacking
    bits_per_index: int


class PaletteImageError(ValueError):
    pass


def _bit_width(palette_count: int) -> int:
    if not 1 <= palette_count <= MAX_PALETTE_COLORS:
        raise PaletteImageError("palette color count is out of range")
    # Public quantized-image grammar stores indices at discrete widths; the
    # 105-color public Timac fixture confirms 17..256 colors use whole bytes.
    for bits in (1, 2, 4, 8):
        if palette_count <= (1 << bits):
            return bits
    return 12


def parse_theme_pixel_rendition(raw: bytes | bytearray | memoryview) -> ThemePixelRendition:
    data = bytes(raw)
    if len(data) < 16:
        raise PaletteImageError("theme pixel rendition is truncated")
    if data[:4] == b"MLEC":
        order = "<"
    elif data[:4] == b"CELM":
        order = ">"
    else:
        raise PaletteImageError("invalid theme pixel rendition magic")
    version, compression_type, raw_length = struct.unpack_from(order + "3I", data, 4)
    if raw_length > len(data) - 16:
        raise PaletteImageError("theme pixel rendition raw payload is truncated")
    return ThemePixelRendition(version, compression_type, data[16:16 + raw_length])


def _unpack_row_indices(data: bytes, width: int, bits_per_index: int) -> bytes:
    row_bytes = (width * bits_per_index + 7) // 8
    if len(data) % row_bytes:
        raise PaletteImageError("quantized index plane has an invalid row length")
    out = bytearray()
    mask = (1 << bits_per_index) - 1
    for row in range(0, len(data), row_bytes):
        buf = int.from_bytes(data[row:row + row_bytes], "big")
        total_bits = row_bytes * 8
        for x in range(width):
            shift = total_bits - bits_per_index * (x + 1)
            out.append((buf >> shift) & mask)
    return bytes(out)


def _pack_row_indices(indices: bytes, width: int, bits_per_index: int) -> bytes:
    row_bytes = (width * bits_per_index + 7) // 8
    if len(indices) % width:
        raise PaletteImageError("unpacked quantized indices are not width-aligned")
    out = bytearray()
    for row in range(0, len(indices), width):
        buf = 0
        total_bits = row_bytes * 8
        for x, value in enumerate(indices[row:row + width]):
            if value >= (1 << bits_per_index):
                raise PaletteImageError("quantized index exceeds bit width")
            shift = total_bits - bits_per_index * (x + 1)
            buf |= int(value) << shift
        out += buf.to_bytes(row_bytes, "big")
    return bytes(out)


def decode_quantized_image_payload(raw_data: bytes, *, width: int, height: int, pixel_format: str = "ARGB") -> QuantizedImageData:
    from . import lzfse_compat
    if pixel_format not in ("ARGB", "RGBW"):
        raise PaletteImageError(f"unsupported quantized pixel format: {pixel_format}")
    decoded = lzfse_compat.decompress(bytes(raw_data))
    if len(decoded) < 10:
        raise PaletteImageError("decoded quantized payload is truncated")
    magic, version = struct.unpack_from("<2I", decoded, 0)
    if magic != MAGIC:
        raise PaletteImageError("decoded quantized payload has an invalid magic")
    if version >= 2:
        raise PaletteImageError("decoded quantized payload version is unsupported")
    palette_count = struct.unpack_from("<H", decoded, 8)[0]
    if not 1 <= palette_count <= MAX_PALETTE_COLORS:
        raise PaletteImageError("decoded quantized payload palette count is invalid")
    entry_size = 8 if pixel_format == "RGBW" else 4
    palette_size = palette_count * entry_size
    palette_start = 10
    palette_end = palette_start + palette_size
    if palette_end > len(decoded):
        raise PaletteImageError("decoded quantized payload palette is truncated")
    bits = _bit_width(palette_count)
    index_plane = decoded[palette_end:]
    row_bytes = (width * bits + 7) // 8
    expected = row_bytes * height
    if len(index_plane) != expected:
        raise PaletteImageError("decoded quantized payload index plane length is invalid")
    palette = tuple(decoded[palette_start + i:palette_start + i + entry_size] for i in range(0, palette_size, entry_size))
    return QuantizedImageData(version, palette, _unpack_row_indices(index_plane, width, bits), bits)


def encode_quantized_image_payload(palette_argb: bytes, indices: bytes, *, width: int, height: int, version: int = 1) -> bytes:
    from . import lzfse_compat
    if version not in (0, 1):
        raise PaletteImageError("quantized payload version must be 0 or 1")
    if len(palette_argb) % 4:
        raise PaletteImageError("palette ARGB buffer length must be divisible by 4")
    palette_count = len(palette_argb) // 4
    bits = _bit_width(palette_count)
    if len(indices) != width * height:
        raise PaletteImageError("quantized index count does not match width*height")
    if any(index >= palette_count for index in indices):
        raise PaletteImageError("quantized index references a missing palette entry")
    packed_indices = _pack_row_indices(indices, width, bits)
    payload = struct.pack("<2IH", MAGIC, version, palette_count) + bytes(palette_argb) + packed_indices
    return lzfse_compat.compress(payload)


def build_palette_img_wrapper(palette_argb: bytes, indices: bytes, *, width: int, height: int, version: int = 1) -> bytes:
    compressed = encode_quantized_image_payload(palette_argb, indices, width=width, height=height, version=version)
    return b"MLEC" + struct.pack("<3I", 0, 8, len(compressed)) + compressed

File: actool/src/codecs/paletteimg.rs
use crate::lzfse;
use byteorder::{LittleEndian, WriteBytesExt};
use thiserror::Error;

pub const MAGIC: u32 = 0xCAFEF00D;
pub const MAX_PALETTE_COLORS: usize = 0x1000;

#[derive(Error, Debug)]
pub enum PaletteImageError {
    #[error("Theme pixel rendition is truncated")]
    Truncated,
    #[error("Invalid theme pixel rendition magic")]
    InvalidMagic,
    #[error("Palette color count out of range: {0}")]
    OutOfRange(usize),
    #[error("Invalid row index encoding")]
    InvalidRowLength,
    #[error("Quantized index exceeds bit width")]
    IndexExceedsWidth,
}

#[derive(Debug, Clone)]
pub struct ThemePixelRendition {
    pub version: u32,
    pub compression_type: u32,
    pub raw_data: Vec<u8>,
}

#[derive(Debug, Clone)]
pub struct QuantizedImageData {
    pub version: u32,
    pub palette: Vec<Vec<u8>>,
    pub indices: Vec<u8>,
    pub bits_per_index: usize,
}

pub fn bit_width(palette_count: usize) -> Result<usize, PaletteImageError> {
    if !(1..=MAX_PALETTE_COLORS).contains(&palette_count) {
        return Err(PaletteImageError::OutOfRange(palette_count));
    }
    for bits in [1, 2, 4, 8] {
        if palette_count <= (1 << bits) {
            return Ok(bits);
        }
    }
    Ok(12)
}

pub fn parse_theme_pixel_rendition(data: &[u8]) -> Result<ThemePixelRendition, PaletteImageError> {
    if data.len() < 16 {
        return Err(PaletteImageError::Truncated);
    }

    let magic = &data[0..4];
    if magic != b"MLEC" && magic != b"CELM" {
        return Err(PaletteImageError::InvalidMagic);
    }

    let is_little = magic == b"MLEC";

    let version = if is_little {
        u32::from_le_bytes(data[4..8].try_into().unwrap())
    } else {
        u32::from_be_bytes(data[4..8].try_into().unwrap())
    };

    let compression_type = if is_little {
        u32::from_le_bytes(data[8..12].try_into().unwrap())
    } else {
        u32::from_be_bytes(data[8..12].try_into().unwrap())
    };

    let raw_data = data[16..].to_vec();

    Ok(ThemePixelRendition {
        version,
        compression_type,
        raw_data,
    })
}

pub fn unpack_row_indices(
    data: &[u8],
    width: usize,
    bits_per_index: usize,
) -> Result<Vec<u8>, PaletteImageError> {
    if width == 0 || bits_per_index == 0 {
        return Ok(Vec::new());
    }
    let row_bytes = (width * bits_per_index + 7) / 8;
    if row_bytes == 0 || data.len() % row_bytes != 0 {
        return Err(PaletteImageError::InvalidRowLength);
    }

    let mut out = Vec::new();
    let mask = (1 << bits_per_index) - 1;

    for row_chunk in data.chunks_exact(row_bytes) {
        let mut buf = 0u128;
        for &b in row_chunk {
            buf = (buf << 8) | (b as u128);
        }
        let total_bits = row_bytes * 8;
        for x in 0..width {
            let shift = total_bits.saturating_sub(bits_per_index * (x + 1));
            out.push(((buf >> shift) as usize & mask) as u8);
        }
    }

    Ok(out)
}

pub fn pack_row_indices(
    indices: &[u8],
    width: usize,
    bits_per_index: usize,
) -> Result<Vec<u8>, PaletteImageError> {
    if width == 0 || bits_per_index == 0 {
        return Ok(Vec::new());
    }
    let row_bytes = (width * bits_per_index + 7) / 8;
    if indices.len() % width != 0 {
        return Err(PaletteImageError::InvalidRowLength);
    }

    let mut out = Vec::new();
    for row_chunk in indices.chunks_exact(width) {
        let mut buf = 0u128;
        let total_bits = row_bytes * 8;
        for (x, &value) in row_chunk.iter().enumerate() {
            if (value as usize) >= (1 << bits_per_index) {
                return Err(PaletteImageError::IndexExceedsWidth);
            }
            let shift = total_bits.saturating_sub(bits_per_index * (x + 1));
            buf |= (value as u128) << shift;
        }

        let mut row_out = vec![0u8; row_bytes];
        for i in (0..row_bytes).rev() {
            row_out[i] = (buf & 0xFF) as u8;
            buf >>= 8;
        }
        out.extend_from_slice(&row_out);
    }

    Ok(out)
}

pub fn decode_quantized_image_payload(
    raw_data: &[u8],
    width: usize,
    _height: usize,
) -> Result<QuantizedImageData, PaletteImageError> {
    let decoded = lzfse::decompress(raw_data).map_err(|_| PaletteImageError::Truncated)?;
    if decoded.len() < 10 {
        return Err(PaletteImageError::Truncated);
    }

    let magic = u32::from_le_bytes(decoded[0..4].try_into().unwrap());
    let version = u32::from_le_bytes(decoded[4..8].try_into().unwrap());
    if magic != MAGIC {
        return Err(PaletteImageError::InvalidMagic);
    }

    let palette_count = u16::from_le_bytes(decoded[8..10].try_into().unwrap()) as usize;
    let bits = bit_width(palette_count)?;

    let palette_start = 10;
    let palette_size = palette_count * 4;
    let palette_end = palette_start + palette_size;

    if palette_end > decoded.len() {
        return Err(PaletteImageError::Truncated);
    }

    let palette: Vec<Vec<u8>> = decoded[palette_start..palette_end]
        .chunks_exact(4)
        .map(|c| c.to_vec())
        .collect();

    let index_plane = &decoded[palette_end..];
    let unpacked_indices = unpack_row_indices(index_plane, width, bits)?;

    Ok(QuantizedImageData {
        version,
        palette,
        indices: unpacked_indices,
        bits_per_index: bits,
    })
}

pub fn encode_quantized_image_payload(
    palette_argb: &[u8],
    indices: &[u8],
    width: usize,
    _height: usize,
) -> Result<Vec<u8>, PaletteImageError> {
    let palette_count = palette_argb.len() / 4;
    let bits = bit_width(palette_count)?;
    let packed_indices = pack_row_indices(indices, width, bits)?;

    let mut payload = Vec::new();
    let _ = payload.write_u32::<LittleEndian>(MAGIC);
    let _ = payload.write_u32::<LittleEndian>(1); // Version 1
    let _ = payload.write_u16::<LittleEndian>(palette_count as u16);
    payload.extend_from_slice(palette_argb);
    payload.extend_from_slice(&packed_indices);

    Ok(lzfse::compress(&payload))
}

pub fn build_palette_img_wrapper(
    palette_argb: &[u8],
    indices: &[u8],
    width: usize,
    height: usize,
) -> Result<Vec<u8>, PaletteImageError> {
    let compressed = encode_quantized_image_payload(palette_argb, indices, width, height)?;
    let mut out = Vec::new();
    out.extend_from_slice(b"MLEC");
    let _ = out.write_u32::<LittleEndian>(0);
    let _ = out.write_u32::<LittleEndian>(8); // Codec 8
    let _ = out.write_u32::<LittleEndian>(compressed.len() as u32);
    out.extend_from_slice(&compressed);

    Ok(out)
}

// --- Auto-generated 1:1 definition shims ---

pub fn _bit_width() {} // Alias for bit_width

pub fn _unpack_row_indices() {} // Alias for unpack_row_indices

pub fn _pack_row_indices() {} // Alias for pack_row_indices
